activerecord/lib/active_record/future_result.rb

```diff
@@ -1,6 +1,76 @@
 # frozen_string_literal: true
 
 module ActiveRecord
+  class Promise < BasicObject
```

Comment:
The public methods in here should be documented - I know this is WIP but I didn't want to forget to say that later when it's no longer draft.